server/groundstation.py

```diff
@@ -0,0 +1,97 @@
+import json
+import time
+from threading import Thread
+
+from interop_handler import InteropHandler
+from uav_handler import UAVHandler
+from dummy_uav_handler import DummyUAVHandler
+
+
+class GroundStation:
+
+    def __init__(self, socket):
+        with open("config.json", "r") as file:
+            self.config = json.load(file)
+
+        self.interop = InteropHandler(self, config=self.config)
+        self.interop_telem_thread = self.update_thread = None
+
+        dummy_uav = self.config["uav"]["dummy"]
+
+        if dummy_uav:
+            self.uav = DummyUAVHandler(self, self.config, socket)
+        else:
+            self.uav = UAVHandler(self, self.config, socket)
+
+        self.interop.login()
+
+        self.func_map = {
+            "i_login": self.interop.login,
+            "i_data": self.interop.get_data,  # key
+            "i_telemetry": self.interop.get_telemetry,
+
+            "i_odlcget": self.interop.odlc_get_queue,  # filter_val
+            "i_odlcadd": self.interop.odlc_add_to_queue,  # 8/9
+            "i_odlcedit": self.interop.odlc_edit,  # 1/10
+            "i_odlcreject": self.interop.odlc_reject,  # id_
+            "i_odlcsubmit": self.interop.odlc_submit,  # id_
+            "i_odlcsave": self.interop.odlc_save_queue,  # filename
+            "i_odlcload": self.interop.odlc_load_queue,  # filename
+
+            "i_mapadd": self.interop.map_add,  # name, image
+            "i_mapsubmit": self.interop.map_submit,  # name
+
+            "m_connect": self.uav.connect,
+            "m_update": self.uav.update,
+            "m_quick": self.uav.quick,
+            "m_stats": self.uav.stats,
+
+            "m_getflightmode": self.uav.get_flight_mode,
+            "m_setflightmode": self.uav.set_flight_mode,  # flightmode
+
+            "m_getparam": self.uav.get_param,  # key
+            "m_setparam": self.uav.set_param,  # key, value
+            "m_getparams": self.uav.get_params,
+            "m_setparams": self.uav.set_params,  # kwargs
+            "m_saveparams": self.uav.save_params,
+            "m_loadparams": self.uav.load_params,
+
+            "m_getcommands": self.uav.get_commands,
+            "m_insertcommand": self.uav.insert_command,  # command, lat, lon, alt
+            "m_clearcommands": self.uav.clear_mission,
+
+            "m_getarmed": self.uav.get_armed,
+            "m_arm": self.uav.arm,
+            "m_disarm": self.uav.disarm
+        }
```

Comment:
Personally I'm not a fan of this method mapping. As I mentioned, it may be better to make a bunch of getters and setters within the main application class, or personally better yet, just let the two classes call each other via `gs.interop.get_data()` or `gs.uav.get_data()`.

Reply:
I still believe that this method is better or long-term debugging and other people working on it, since all functions and their maps are defined here. If anything needs to be modified, it can be done in this one dictionary and everything will still run smoothly. We can discuss this more later.

Reply:
That same argument could be made about the whole class itself. "If anything needs to be modified, it can be done with this one [class] and everything will run smoothly." I think it adds an extra step, to be honest. At first I kind of agreed because I felt like something like `gs.uav.get_data()` broke encapsulation, but then I felt like it was fine because it's just calling a submodule (and that's not a great explanation. Not sure how to say it).

Reply:
My problem with this is of unknown errors, namely CircularImport-like errors. For example, if you import a file that imports the original file itself, you get a CircularImport exception. I'm not sure how reasonable it would be to directly reference a class that calls the same function in the main class, and I believe some classes do call functions in the other class, which call the original class back. I can definitely try this, but I feel like it's not worth the risk.

Reply:
Well it wouldn't be a circular import because either file is not importing the other. Only the main ground station file is importing both. As for a circular reference, I don't think it would be a problem because at the core of it, it's the main ground station file holding a reference to the two files. As long as UAV and Interop don't hold a reference to each other, it would be fine (and they don't; at least, not directly). Again, only the main ground station class holds the references.

Reply:
Or well the whole thing would be flawed from the start because for both UAV and Interop, they hold a reference to GS, which in turn holds a reference back. It's not a reference to each other that's the problem, it's just the inherent structure of it. However, I'm not sure if it would be a problem. Circular references if I remember correctly are only a problem when the objects are garbage corrected. In theory, the garbage collection for them for only happens at the termination of the program, so it would not be a problem. Anyways, it just makes the garbage collection slow, or at least that's what I'm reading. You should probably look into it and see if it's OK. Relevant stackoverflow: https://stackoverflow.com/questions/2428301/should-i-worry-about-circular-references-in-python.